### PR TITLE
feat: (WIP) started conversion of tapable to TypeScript

### PR DESCRIPTION
**Pull request by TheLarkInn**
Still very work in progress. One thing to note, is that TS doesn't support `..args` that are not the final parameter. There is an ugly workaround that would have to be considered. https://github.com/Microsoft/TypeScript/issues/7347

### CODE REVIEW

**Comment by jayphelps** [comment]
not ideal, but the API could change to be Promise-based instead of callback based.

```ts
applyPluginsAsyncSeries(name: string, ...args: Array<any>): PromiseLike
```

That's probably an entire bikeshedding session on its own though...

/ducks out before someone throws a shoe

**Comment by TheLarkInn** [comment]
I mean promises would be really cool is just that it would be a ginourmous breaking change.

**Comment by e-cloud** [comment]
promise is good. Breaking change is inevitable when introducing awesome feature in most of time.

As webpack 2 will be incompatible with webpack 1. Introducing breaking change would affect webpack 2 basically

**Comment by e-cloud** [comment]
One thing worth mentioned is documentation. Especially, the `apply` method overlay the built-in one worth documenting why doing this, what this do, how to use it.

As webpack grows so popular, the core concept is one of the most important thing to be clarified, for further improvement.

File: lib/Tapable.ts [hunk 1]
```diff
@@ -2,40 +2,120 @@
...
+		if(!this._plugins[name]) return init;
+
+		let plugins = this._plugins[name];
+		let current = init;
+		for(let i = 0; i < plugins.length; i++)
+			current = plugins[i].apply(this, [current].concat(args));
+		return current;
+	}
+
+	protected applyPluginsWaterfall0(name: string, init: any) {
```
Comment (TheLarkInn) [inline lib/Tapable.ts]: So changing the actual name of the function would require a huge changes to webpack/core, but once I've gotten this fully migrated, I can add a comment to explain how some are ussing `call` and some use `apply` thus the strange name.

**Comment by TheLarkInn** [comment]
I agree with you. Most users don't even understand Tapable or the entire webpack internals. It certainly is an opprotunity to fix that through code quality.

**Comment by TheLarkInn** [comment]
Types are on hold at the moment. Would require potentially a design change especially with delegation.

**Pull request closed without merging**